### PR TITLE
Add Qt olm binding

### PR DESCRIPTION
**Pull request by ognarb**
Start working on better Qt olm binding. The api and implementation is a plain port of the rust lib: olm-rs.

### CODE REVIEW

File: lib/olm/e2ee.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+struct IdentityKeys
+{
+    QByteArray curve25519;
+    QByteArray ed25519;
+};
+
+//! Struct representing the the one-time keys.
```
Comment (TobiasFella) [inline lib/olm/e2ee.h:57]: ``
```suggestion
//! Struct representing the one-time keys.
```

File: lib/olm/qolmaccount.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    }
+    return keys[keyType];
+}
+
+bool operator==(const IdentityKeys& lhs, const IdentityKeys& rhs)
+{
+    return lhs.curve25519 == rhs.curve25519 &&& lhs.ed25519 == rhs.ed25519;
+}
+
+// Conver olm error to enum
```
Comment (TobiasFella) [inline lib/olm/qolmaccount.cpp:33]:
```suggestion
// Convert olm error to enum
```

File: lib/olm/qolmoutboundsession.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+class QOlmOutboundGroupSession
+{
+public:
+    ~QOlmOutboundGroupSession();
+    //! Creates a new instance of `QOlmOutboundGroupSession`.
+    static std::variant<QOlmOutboundGroupSession, OlmError> create();
+    //! Serialises an `QOlmOutboundGroupSession` to encrypted Base64.
+    std::variant<QByteArray, OlmError> pickle(const PicklingMode &mode);
+    //! Deserialises from encrypted Base64 that was previously obtained by
+    //! pickling an `QOlmOutboundGroupSession`.
```
Comment (TobiasFella) [inline lib/olm/qolmoutboundsession.h:23]:
```suggestion
    //! pickling a `QOlmOutboundGroupSession`.
```

**Comment by TobiasFella** [comment]
Fails to build when building tests but not E2EE

**Comment by ognarb** [comment]
> Fails to build when building tests but not E2EE

This should now be fixed. I just needed to ifdef everything. A good goal would still be at some point to remove all this ifdef once this feature is stable and don't crash :)

File: autotests/testgroupsession.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    const auto ogsId = std::get<QByteArray>(ogs.sessionId());
+    QVERIFY(QByteArray::fromBase64Encoding(ogsId).decodingStatus == QByteArray::Base64DecodingStatus::Ok);
+    QCOMPARE(0, ogs.sessionMessageIndex());
+
+    auto ogsPickled = std::get<QByteArray>(ogs.pickle(Unencrypted {}));
+    ogs = std::get<QOlmOutboundGroupSession>(QOlmOutboundGroupSession::unpickle(ogsPickled, Unencrypted {}));
+    QCOMPARE(ogsId, std::get<QByteArray>(ogs.sessionId()));
+
+    qDebug() << std::get<QByteArray>(ogs.sessionKey());
+    auto igs = std::get<QOlmInboundGroupSession>(QOlmInboundGroupSession::create(std::get<QByteArray>(ogs.sessionKey())));
```
Comment (ognarb) [inline autotests/testgroupsession.cpp:25]: It's crashing here with a BAD_SIGNATURE error and I'm not sure why?

**Comment by KitsuneRal** [comment]
I feel a bit uneasy about having `olm/` directory inside libQuotient tree - it's not immediately clear whether a given header file comes from Olm itself or the wrapper. Perhaps rename it to `e2ee/` or `qolm`?

Also: did I get it right that you plan to eventually get rid of QtOlm dependency, rather than rewrite it?

**Comment by ognarb** [comment]
> I feel a bit uneasy about having `olm/` directory inside libQuotient tree - it's not immediately clear whether a given header file comes from Olm itself or the wrapper. Perhaps rename it to `e2ee/` or `qolm`?

Yes, it's probably that I will do soon, even more as I'm adding stuff related to encryption but not to olm specific (e.g. key export)

> Also: did I get it right that you plan to eventually get rid of QtOlm dependency, rather than rewrite it?

Yes, the goal is to get rid of QtOlm. For the moment, I still have dependencies to QtOlm but it's only to make sure the code still compiles but I think I will be soon at a point where I can replace it with the new API.

File: lib/converters.cpp [hunk 5]
```diff
@@ -3,17 +3,28 @@
...
+    if (v.canConvert<SignedOneTimeKey>()) {
+        return toJson(v.value<SignedOneTimeKey>());
+    }
     return QJsonValue::fromVariant(v);
 }
 
 QVariant JsonConverter<QVariant>::load(const QJsonValue& jv)
 {
+    if (jv.isObject()) {
+        QJsonObject obj = jv.toObject();
```
Comment (nicolasfella) [inline lib/converters.cpp:22]:
```suggestion
        const QJsonObject obj = jv.toObject();
```

File: lib/converters.cpp [hunk 6]
```diff
@@ -3,17 +3,28 @@
...
+        if (obj.contains("key") && obj.contains("signatures")) {
```
Comment (nicolasfella) [inline lib/converters.cpp:23]:
```suggestion
        if (obj.contains(QLatin1String("key")) && obj.contains(QLatin1String("signatures"))) {
```

File: autotests/testolmaccount.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    class Connection;
+}
+
+class TestOlmAccount : public QObject
+{
+    Q_OBJECT
+
+
+private Q_SLOTS:
+    void pickleUnpickedTest();
```
Comment (TobiasFella) [inline autotests/testolmaccount.h:18]: pickleUnpickled?

File: lib/converters.cpp [hunk 8]
```diff
@@ -3,17 +3,28 @@
...
+            SignedOneTimeKey signedOneTimeKeys;
+            signedOneTimeKeys.key = obj["key"].toString();
```
Comment (nicolasfella) [inline lib/converters.cpp:25]:
```suggestion
            signedOneTimeKeys.key = obj[QLatin1String("key")].toString();
```

File: autotests/testolmaccount.cpp [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+    olmAccount.generateOneTimeKeys(20);
+    const auto oneTimeKeysFilled = olmAccount.oneTimeKeys();
+    QCOMPARE(20, oneTimeKeysFilled.curve25519().count());
+}
+
+void TestOlmAccount::deviceKeys()
+{
+    // copied from mtxclient
+    DeviceKeys device1;
+    device1.userId   = "@alice:example.com";
```
Comment (TobiasFella) [inline autotests/testolmaccount.cpp:76]:
```suggestion
    device1.userId = "@alice:example.com";
```

File: lib/crypto/qolmaccount.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    //! Creates an outbound session for sending messages to a specific
+    /// identity and one time key.
+    std::variant<std::unique_ptr<QOlmSession>, QOlmError> createOutboundSession(const QByteArray &theirIdentityKey, const QByteArray &theirOneTimeKey);
+
+    // HACK do not use directly
+    QOlmAccount(OlmAccount *account);
+    OlmAccount *data();
+private:
+    OlmAccount *m_account = nullptr; // owning
```
Comment (nicolasfella) [inline lib/crypto/qolmaccount.h:99]: unique_ptr?

File: autotests/testolmaccount.cpp [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+
+void TestOlmAccount::uploadIdentityKey()
+{
+    CREATE_CONNECTION(conn, "alice", "secret", "AlicePhone")
+
+    auto olmAccount = conn->olmAccount();
+    auto idKeys = olmAccount->identityKeys();
+
+    QVERIFY(idKeys.curve25519.size() > 10);
+    QVERIFY(idKeys.curve25519.size() > 10);
```
Comment (TobiasFella) [inline autotests/testolmaccount.cpp:191]: duplicate line?

File: lib/room.cpp [hunk 12]
```diff
@@ -62,13 +62,12 @@
 #include <functional>
 
 #ifdef Quotient_E2EE_ENABLED
-#include <account.h> // QtOlm
-#include <errors.h> // QtOlm
-#include <groupsession.h> // QtOlm
+#    include "crypto/qolmaccount.h"
```
Comment (nicolasfella) [inline lib/room.cpp:65]:
```suggestion
#include "crypto/qolmaccount.h"
```

File: lib/networkaccessmanager.cpp [hunk 13]
```diff
@@ -15,7 +15,12 @@ class NetworkAccessManager::Private {
 
 NetworkAccessManager::NetworkAccessManager(QObject* parent)
     : QNetworkAccessManager(parent), d(std::make_unique<Private>())
-{}
+{
+    connect(this, &QNetworkAccessManager::sslErrors, this, [](QNetworkReply *reply, const QList<QSslError> &errors)
```
Comment (nicolasfella) [inline lib/networkaccessmanager.cpp:19]: Is this related?

File: lib/encryptionmanager.cpp [hunk 14]
```diff
@@ -328,22 +333,22 @@ QString EncryptionManager::sessionDecryptMessage(
...
     if (type == 0) {
-        PreKeyMessage preKeyMessage { body };
-        decrypted = d->sessionDecrypt(reinterpret_cast<Message*>(&preKeyMessage),
-                                      senderKey);
+        QOlmMessage preKeyMessage = QOlmMessage(body, QOlmMessage::PreKey);
+        decrypted = d->sessionDecrypt(preKeyMessage, senderKey);
     } else if (type == 1) {
-        Message message { body };
-        decrypted = d->sessionDecrypt(&message, senderKey);
+        QOlmMessage message = QOlmMessage(body, QOlmMessage::PreKey);
```
Comment (nicolasfella) [inline lib/encryptionmanager.cpp:339]:
```suggestion
        QOlmMessage message(body, QOlmMessage::PreKey);
```

File: lib/encryptionmanager.cpp [hunk 15]
```diff
@@ -328,22 +333,22 @@ QString EncryptionManager::sessionDecryptMessage(
     int type = personalCipherObject.value(TypeKeyL).toInt(-1);
     QByteArray body = personalCipherObject.value(BodyKeyL).toString().toLatin1();
     if (type == 0) {
-        PreKeyMessage preKeyMessage { body };
-        decrypted = d->sessionDecrypt(reinterpret_cast<Message*>(&preKeyMessage),
-                                      senderKey);
+        QOlmMessage preKeyMessage = QOlmMessage(body, QOlmMessage::PreKey);
```
Comment (nicolasfella) [inline lib/encryptionmanager.cpp:336]:
```suggestion
        QOlmMessage preKeyMessage(body, QOlmMessage::PreKey);
```

File: lib/encryptionmanager.cpp [hunk 16]
```diff
@@ -72,7 +69,7 @@ class EncryptionManager::Private {
     UploadKeysJob* uploadOneTimeKeysJob = nullptr;
     QueryKeysJob* queryKeysJob = nullptr;
 
-    QScopedPointer<Account> olmAccount;
+    QScopedPointer<QOlmAccount> olmAccount;
```
Comment (nicolasfella) [inline lib/encryptionmanager.cpp:72]: unique_ptr for consistency?

File: lib/crypto/qolmutils.cpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+#include "crypto/qolmutils.h"
+#include <QDebug>
+#include <openssl/rand.h>
+
+using namespace Quotient;
+
+QByteArray Quotient::toKey(const Quotient::PicklingMode &mode)
+{
+    if (std::holds_alternative<Quotient::Unencrypted>(mode)) {
+        return "";
```
Comment (nicolasfella) [inline lib/crypto/qolmutils.cpp:15]:
```suggestion
        return {};
```

File: lib/networkaccessmanager.cpp (hunk 13 above)
Comment (ognarb) [reply lib/networkaccessmanager.cpp:19]: it's for the test but need a better solution because we can't ignore the SSL errors in production...

File: lib/crypto/qolmaccount.h (hunk 10 above)
Comment (ognarb) [reply lib/crypto/qolmaccount.h:99]: An unique_ptr with its own custom deleter could work but it's more just add more complexity.

File: lib/networkaccessmanager.cpp (hunk 13 above)
Comment (ognarb) [reply lib/networkaccessmanager.cpp:19]: Patch that should fix it but doesn't work. Probably because of a stupid error somewhere.

```
diff --git a/autotests/testolmaccount.cpp b/autotests/testolmaccount.cpp
index 9134224..eb44791 100644
--- a/autotests/testolmaccount.cpp
+++ b/autotests/testolmaccount.cpp
@@ -165,6 +165,7 @@ void TestOlmAccount::encryptedFile()
 
 #define CREATE_CONNECTION(VAR, USERNAME, SECRET, DEVICE_NAME) \
     auto VAR = std::make_shared<Connection>(); \
+    VAR->ignoreSslErrors(true); \
     (VAR) ->resolveServer("@alice:localhost:" + QString::number(443)); \
     connect( (VAR) .get(), &Connection::loginFlowsChanged, this, [this, VAR ] () { \
         (VAR) ->loginWithPassword( (USERNAME) , SECRET , DEVICE_NAME , ""); \
diff --git a/lib/connection.cpp b/lib/connection.cpp
index 070aac0..d01aab3 100644
--- a/lib/connection.cpp
+++ b/lib/connection.cpp
@@ -573,6 +573,11 @@ void Connection::sync(int timeout)
     });
 }
 
+void Connection::ignoreSslErrors(bool ignore)
+{
+    connectionData()->ignoreSslErrors(ignore);
+}
+
 void Connection::syncLoop(int timeout)
 {
     if (d->syncLoopConnection && d->syncTimeout == timeout) {
diff --git a/lib/connection.h b/lib/connection.h
index 9a952e0..f3a6b8c 100644
--- a/lib/connection.h
+++ b/lib/connection.h
@@ -481,6 +481,11 @@ public:
         setUserFactory(defaultUserFactory<T>());
     }
 
+    /// Ignore ssl errors (usefull for automated testing with local synapse
+    /// instance).
+    /// \internal
+    void ignoreSslErrors(bool ignore);
+
 public Q_SLOTS:
     /** Set the homeserver base URL */
     void setHomeserver(const QUrl& baseUrl);
diff --git a/lib/connectiondata.cpp b/lib/connectiondata.cpp
index e54d909..123febe 100644
--- a/lib/connectiondata.cpp
+++ b/lib/connectiondata.cpp
@@ -140,6 +140,12 @@ bool ConnectionData::needsToken(const QString& requestName) const
            != d->needToken.cend();
 }
 
+void ConnectionData::ignoreSslErrors(bool ignore) const
+{
+    auto quotientNam = static_cast<NetworkAccessManager>(nam());
+    quotientNam.ignoreSslErrors(ignore);
+}
+
 void ConnectionData::setDeviceId(const QString& deviceId)
 {
     d->deviceId = deviceId;
diff --git a/lib/connectiondata.h b/lib/connectiondata.h
index 7dd96f2..00d056e 100644
--- a/lib/connectiondata.h
+++ b/lib/connectiondata.h
@@ -29,6 +29,7 @@ public:
     bool needsToken(const QString& requestName) const;
     QNetworkAccessManager* nam() const;
 
+    void ignoreSslErrors(bool ignore = true) const;
     void setBaseUrl(QUrl baseUrl);
     void setToken(QByteArray accessToken);
     [[deprecated("Use setBaseUrl() instead")]]
diff --git a/lib/networkaccessmanager.cpp b/lib/networkaccessmanager.cpp
index c6e761e..1276305 100644
--- a/lib/networkaccessmanager.cpp
+++ b/lib/networkaccessmanager.cpp
@@ -16,10 +16,6 @@ public:
 NetworkAccessManager::NetworkAccessManager(QObject* parent)
     : QNetworkAccessManager(parent), d(std::make_unique<Private>())
 {
-    connect(this, &QNetworkAccessManager::sslErrors, this, [](QNetworkReply *reply, const QList<QSslError> &errors)
-            {
-                reply->ignoreSslErrors();
-            });
 }
 
 QList<QSslError> NetworkAccessManager::ignoredSslErrors() const
@@ -27,6 +23,19 @@ QList<QSslError> NetworkAccessManager::ignoredSslErrors() const
     return d->ignoredSslErrors;
 }
 
+void NetworkAccessManager::ignoreSslErrors(bool ignore) const
+{
+    if (ignore) {
+        connect(this, &QNetworkAccessManager::sslErrors, this, [](QNetworkReply *reply, const QList<QSslError> &errors)
+                {
+    qDebug() << 24 / 0;
+                    reply->ignoreSslErrors();
+                });
+    } else {
+        disconnect(this, &QNetworkAccessManager::sslErrors, this, nullptr);
+    }
+}
+
 void NetworkAccessManager::addIgnoredSslError(const QSslError& error)
 {
     d->ignoredSslErrors << error;
diff --git a/lib/networkaccessmanager.h b/lib/networkaccessmanager.h
index 47729a1..22280e0 100644
--- a/lib/networkaccessmanager.h
+++ b/lib/networkaccessmanager.h
@@ -17,6 +17,7 @@ public:
     QList<QSslError> ignoredSslErrors() const;
     void addIgnoredSslError(const QSslError& error);
     void clearIgnoredSslErrors();
+    void ignoreSslErrors(bool ignore = true) const;
 
     /** Get a pointer to the singleton */
     static NetworkAccessManager* instance();
```

File: lib/crypto/qolmaccount.cpp [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+        return new UploadKeysJob(deviceKeys);
+    }
+
+    // Sign & append the one time keys.
+    auto temp = signOneTimeKeys(oneTimeKeys);
+    QHash<QString, QVariant> oneTimeKeysSigned;
+    for (const auto &[keyId, key] : asKeyValueRange(temp)) {
+        QVariant keyVar;
+        keyVar.setValue(key);
+        oneTimeKeysSigned[keyId] = keyVar;
```
Comment (TobiasFella) [inline lib/crypto/qolmaccount.cpp:259]:
```suggestion
        oneTimeKeysSigned[keyId] = QVariant::fromValue(key);
```

File: lib/crypto/qolmaccount.h [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    //! Creates a new instance of OlmAccount. During the instantiation
+    //! the Ed25519 fingerprint key pair and the Curve25519 identity key
+    //! pair are generated. For more information see <a
+    //! href="https://matrix.org/docs/guides/e2e_implementation.html#keys-used-in-end-to-end-encryption">here</a>.
+    //! This needs to be called before any other action or use unpickle() instead.
+    void createNewAccount();
+
+    //! Deserialises from encrypted Base64 that was previously obtained by pickling a `QOlmAccount`.
+    //! This needs to be called before any other action or use createNewAccount() instead.
+    void unpickle(QByteArray &picked, const PicklingMode &mode);
```
Comment (TobiasFella) [inline lib/crypto/qolmaccount.h:40]:
```suggestion
    void unpickle(QByteArray &pickled, const PicklingMode &mode);
```

File: lib/crypto/qolmaccount.h [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    //! currently hold.
+    size_t maxNumberOfOneTimeKeys() const;
+
+    //! Generates the supplied number of one time keys.
+    size_t generateOneTimeKeys(size_t numberOfKeys) const;
+
+    //! Gets the OlmAccount's one time keys formatted as JSON.
+    OneTimeKeys oneTimeKeys() const;
+
+    //! Sign all time key.
```
Comment (TobiasFella) [inline lib/crypto/qolmaccount.h:65]:
```suggestion
    //! Sign all one time keys.
```

File: autotests/testolmaccount.cpp [hunk 21]
```diff
@@ -0,0 +1,419 @@
...
+
+    QHash<QString, QHash<QString, QString>> oneTimeKeys;
+    for (const auto &d : devices_) {
+        oneTimeKeys[alice->userId()] = QHash<QString, QString>();
+        oneTimeKeys[alice->userId()][d] = SignedCurve25519Key;
+    }
+    auto job = bob->callApi<ClaimKeysJob>(oneTimeKeys);
+    connect(job, &BaseJob::result, this, [bob, job] {
+        const auto userId = bob->userId();
+        const auto deviceId = bob->deviceId();
```
Comment (TobiasFella) [inline autotests/testolmaccount.cpp:412]:
```suggestion
```

File: autotests/testolmaccount.h [hunk 22]
```diff
@@ -0,0 +1,31 @@
+// SPDX-FileCopyrightText: 2021 Carl Schwan <carlschwan@kde.org>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <QtTest/QtTest>
```
Comment (TobiasFella) [inline autotests/testolmaccount.h:5]:
```suggestion
#include <QTest>
```

File: autotests/testolmaccount.h [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Quotient {
+    class Connection;
+}
+
+class TestOlmAccount : public QObject
+{
+    Q_OBJECT
+
+
```
Comment (TobiasFella) [inline autotests/testolmaccount.h:16]:
```suggestion
```

File: autotests/testolmsession.cpp [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    const QByteArray oneTimeKeyA("WzsbsjD85iB1R32iWxfJdwkgmdz29ClMbJSJziECYwk");
+    const QByteArray identityKeyB("q/YhJtog/5VHCAS9rM9uUf6AaFk1yPe4GYuyUOXyQCg");
+    const QByteArray oneTimeKeyB("oWvzryma+B2onYjo3hM6A3Mgo/Yepm8HvgSvwZMTnjQ");
+    auto outbound = std::get<std::unique_ptr<QOlmSession>>(accountA
+        .createOutboundSession(identityKeyB, oneTimeKeyB));
+
+    const auto preKey = outbound->encrypt(""); // Payload does not matter for PreKey
+
+    if (preKey.type() != QOlmMessage::PreKey) {
+        throw "Wrong first message type received, can't create session";
```
Comment (TobiasFella) [inline autotests/testolmsession.cpp:29]:
```suggestion
        QFAIL("Wrong first message type received, can't create session");
```

File: autotests/testolmsession.h [hunk 25]
```diff
@@ -0,0 +1,14 @@
+// SPDX-FileCopyrightText: 2021 Carl Schwan <carlschwan@kde.org>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <QtTest/QtTest>
```
Comment (TobiasFella) [inline autotests/testolmsession.h:5]:
```suggestion
#include <QTest>
```

File: autotests/testolmutility.cpp [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+    auto msgObj = QJsonObject({{"key", firstKey}});
+    auto sig = aliceOlm->sign(msgObj);
+
+    auto msg = QJsonDocument(msgObj).toJson(QJsonDocument::Compact);
+
+    auto utilityBuf = new uint8_t[olm_utility_size()];
+    auto utility = olm_utility(utilityBuf);
+
+
+    QByteArray signatureBuf1(sig.length(), '0');
```
[the remark this answers is not in the document]
Comment (ognarb) [reply autotests/testolmutility.cpp:65]: If I remember correctly olm_ed25519_verify will modify sig and we still need the original sig later.

File: lib/connection.cpp [hunk 27]
```diff
@@ -230,16 +234,17 @@ class Connection::Private {
                 .value(Ed25519Key).toString();
         if (ourKey
             != QString::fromUtf8(
-                encryptionManager->account()->ed25519IdentityKey())) {
+                encryptionManager->account()->identityKeys().ed25519)) {
             qCDebug(E2EE) << "Found key" << ourKey
                           << "instead of ours own ed25519 key"
-                          << encryptionManager->account()->ed25519IdentityKey()
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/connection.cpp:236]: I guess this is just debugging code, for the moment I completely disabled this code since it was crashy

File: lib/converters.cpp [hunk 28]
```diff
@@ -3,17 +3,28 @@
...
+#include <QVariant>
+#include "crypto/e2ee.h"
 
 using namespace Quotient;
 
 QJsonValue JsonConverter<QVariant>::dump(const QVariant& v)
 {
+    if (v.canConvert<SignedOneTimeKey>()) {
+        return toJson(v.value<SignedOneTimeKey>());
+    }
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/converters.cpp:15]: It's a hack, unfortunately, the JSON /C++ bridge can't really handle the case where something is either an object or a string.

File: lib/crypto/qolmaccount.cpp [hunk 29]
```diff
@@ -0,0 +1,322 @@
+// SPDX-FileCopyrightText: 2021 Carl Schwan <carlschwan@kde.org>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#ifdef Quotient_E2EE_ENABLED
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/crypto/qolmaccount.cpp:5]: yes definitely

File: lib/crypto/qolmaccount.cpp [hunk 30]
```diff
@@ -0,0 +1,322 @@
...
+    return keys[keyType];
+}
+
+bool operator==(const IdentityKeys& lhs, const IdentityKeys& rhs)
+{
+    return lhs.curve25519 == rhs.curve25519 &&& lhs.ed25519 == rhs.ed25519;
+}
+
+// Convert olm error to enum
+QOlmError lastError(OlmAccount *account) {
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/crypto/qolmaccount.cpp:37]: it's an internal function, that could be a private method too but I don't really see an advantage.

File: lib/crypto/qolmaccount.cpp [hunk 31]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+void QOlmAccount::createNewAccount()
+{
+    m_account = olm_account(new uint8_t[olm_account_size()]);
+    size_t randomSize = olm_create_account_random_length(m_account);
+    QByteArray randomData = getRandom(randomSize);
+    const auto error = olm_create_account(m_account, randomData.data(), randomSize);
+    if (error == olm_error()) {
+        throw lastError(m_account);
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/crypto/qolmaccount.cpp:69]: Generally I tried to only throw for errors that can't and shouldn't be recovered. In that case, it should only throw when there was a problem when allocating the memory.

File: lib/crypto/qolmaccount.cpp [hunk 32]
```diff
@@ -0,0 +1,322 @@
...
+    QByteArray oneTimeKeysBuffer(oneTimeKeyLength, '0');
+
+    const auto error = olm_account_one_time_keys(m_account, oneTimeKeysBuffer.data(), oneTimeKeyLength);
+    if (error == olm_error()) {
+        throw lastError(m_account);
+    }
+    const auto json = QJsonDocument::fromJson(oneTimeKeysBuffer).object();
+    OneTimeKeys oneTimeKeys;
+
+    for (const QJsonValue &key1 : json.keys()) {
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/crypto/qolmaccount.cpp:177]: We can try :) or at least move that code to converter.h

File: lib/crypto/qolmutility.cpp [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+std::variant<bool, QOlmError> QOlmUtility::ed25519Verify(const QByteArray &key,
+        const QByteArray &message, const QByteArray &signature)
+{
+    QByteArray signatureBuf(signature.length(), '0');
+    std::copy(signature.begin(), signature.end(), signatureBuf.begin());
+
+    const auto ret = olm_ed25519_verify(m_utility, key.data(), key.size(),
+            message.data(), message.size(), (void *)signatureBuf.data(), signatureBuf.size());
+
+    const auto error = ret;
```
[the remark this answers is not in the document]
Comment (ognarb) [reply lib/crypto/qolmutility.cpp:55]: yeah, that stuff should be simplified (same with the if condition underneath)

**Comment by TobiasFella** [comment]
Since there will probably be other E2EE encryption systems in matrix sooner or later, should the olm stuff be moved to crypto/qolm or something?

File: lib/crypto/qolminboundsession.cpp [hunk 34]
```diff
@@ -0,0 +1,155 @@
...
+
+std::unique_ptr<QOlmInboundGroupSession> QOlmInboundGroupSession::create(const QByteArray &key)
+{
+    const auto olmInboundGroupSession = olm_inbound_group_session(new uint8_t[olm_inbound_group_session_size()]);
+    const auto temp = key;
+    const auto error = olm_init_inbound_group_session(olmInboundGroupSession,
+            reinterpret_cast<const uint8_t *>(temp.data()), temp.size());
+
+    if (error == olm_error()) {
+        throw lastError(olmInboundGroupSession);
```
Comment (TobiasFella) [inline lib/crypto/qolminboundsession.cpp:36]: we probably shouldn't fail completely for these errors

File: lib/crypto/qolminboundsession.cpp [hunk 35]
```diff
@@ -0,0 +1,155 @@
...
+
+std::unique_ptr<QOlmInboundGroupSession> QOlmInboundGroupSession::import(const QByteArray &key)
+{
+    const auto olmInboundGroupSession = olm_inbound_group_session(new uint8_t[olm_inbound_group_session_size()]);
+    QByteArray keyBuf = key;
+
+    const auto error = olm_import_inbound_group_session(olmInboundGroupSession,
+            reinterpret_cast<const uint8_t *>(keyBuf.data()), keyBuf.size());
+    if (error == olm_error()) {
+        throw lastError(olmInboundGroupSession);
```
Comment (TobiasFella) [inline lib/crypto/qolminboundsession.cpp:51]: same here

File: lib/crypto/qolmsession.cpp [hunk 36]
```diff
@@ -0,0 +1,256 @@
...
+OlmSession* QOlmSession::create()
+{
+    return olm_session(new uint8_t[olm_session_size()]);
+}
+
+std::variant<std::unique_ptr<QOlmSession>, QOlmError> QOlmSession::createInbound(QOlmAccount *account, const QOlmMessage &preKeyMessage, bool from, const QString &theirIdentityKey)
+{
+    if (preKeyMessage.type() != QOlmMessage::PreKey) {
+        qCDebug(E2EE) << "The message is not a pre-key";
+        throw BadMessageFormat;
```
Comment (TobiasFella) [inline lib/crypto/qolmsession.cpp:34]: Can we assume that this will never happen if the client uses this correctly?

File: lib/crypto/qolmsession.cpp [hunk 37]
```diff
@@ -0,0 +1,256 @@
...
+    std::memcpy(output.data(), plaintextBuf.data(), plaintextResultLen);
+    plaintextBuf.clear();
+    return output;
+}
+
+QOlmMessage::Type QOlmSession::encryptMessageType()
+{
+    const auto messageTypeResult = olm_encrypt_message_type(m_session);
+    if (messageTypeResult == olm_error()) {
+        throw lastError(m_session);
```
Comment (TobiasFella) [inline lib/crypto/qolmsession.cpp:190]: The olm documentation does not specify which errors might happen here. we should look into the implementation to figure out if we should handle them or not

**Comment by TobiasFella** [comment]
we could also assert the unrecoverable errors instead of throwing them, since they should not happen if the program is reasonably correct

**Comment by KitsuneRal** [comment]
Upon a quick look-through, without really analysing the code:
- Please double-check the code is formatted according to `.clang-format`.
- Please indicate the Qt module of the header: `#include <QtCore/QDebug>`, not `#include <QDebug>`.
- I understand the idea of throwing when things are utterly grim and irrecoverable; but in some cases `throw` is combined with the error object, which is very confusing from the API standpoint. Please let's _really_ try to go without exceptions, as long as there's `QOlmError`.

I will look closer into the code; it would help if you could clearly declare the scope of the work accomplished by now (what is supposed to work, namely).

**Comment by ognarb** [comment]
I pushed this to the work/olm branch. I will close this MR and new E2EE related MR should be done against the work/olm branch. Once everything is working we can merge the work/olm branch inside master. :)

**Comment by hugohn** [comment]
Hi guys, is this still ongoing work? was there an MR that replaced this MR?

**Comment by CarlSchwan** [comment]
> Hi guys, is this still ongoing work? was there an MR that replaced this MR?

This was merged in another mr. Message decryption should now work if the compile flag is set

**Comment by hugohn** [comment]
> > Hi guys, is this still ongoing work? was there an MR that replaced this MR?
> 
> This was merged in another mr. Message decryption should now work if the compile flag is set

Awesome! Since 0.7 is not out yet, I assume message decryption is currently available only in the dev branch, correct?

Also do you know when will 0.7 be released?

**Comment by KitsuneRal** [comment]
Hopefully sometime later in the summer, contingent on me having at least some reasonable time for the project (and in any case it likely won't have complete E2EE, rather an early beta of sorts).
